chainindex/events.go

```diff
@@ -0,0 +1,501 @@
+
+			// Insert event entries
+			for _, entry := range event.Entries {
+				_, err := tx.Stmt(si.insertEventEntryStmt).Exec(
+					eventID,
+					isIndexedValue(entry.Flags),
+					[]byte{entry.Flags},
+					entry.Key,
+					entry.Codec,
+					entry.Value,
+				)
+				if err != nil {
+					return xerrors.Errorf("failed to insert event entry: %w", err)
+				}
+			}
+			eventCount++
+		}
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) loadExecutedMessages(ctx context.Context, msgTs, rctTs *types.TipSet) ([]executedMessage, error) {
+	msgs, err := si.cs.MessagesForTipset(ctx, msgTs)
+	if err != nil {
+		return nil, xerrors.Errorf("error getting messages for tipset: %w", err)
+	}
+
+	st := si.cs.ActorStore(ctx)
+
+	receiptsArr, err := blockadt.AsArray(st, rctTs.Blocks()[0].ParentMessageReceipts)
+	if err != nil {
+		return nil, xerrors.Errorf("error loading message receipts array: %w", err)
+	}
+
+	if uint64(len(msgs)) != receiptsArr.Length() {
+		return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), receiptsArr.Length())
+	}
+
+	ems := make([]executedMessage, len(msgs))
+
+	for i := 0; i < len(msgs); i++ {
+		ems[i].msg = msgs[i]
+
+		var rct types.MessageReceipt
+		found, err := receiptsArr.Get(uint64(i), &rct)
+		if err != nil {
+			return nil, xerrors.Errorf("error loading receipt %d: %w", i, err)
+		}
+		if !found {
+			return nil, xerrors.Errorf("receipt %d not found", i)
+		}
+		ems[i].rct = rct
+
+		// no events in the receipt
+		if rct.EventsRoot == nil {
+			continue
+		}
+
+		eventsArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+		if err != nil {
+			return nil, xerrors.Errorf("error loading events amt: %w", err)
+		}
+
+		ems[i].evs = make([]types.Event, eventsArr.Len())
+		var evt types.Event
+		err = eventsArr.ForEach(ctx, func(u uint64, deferred *cbg.Deferred) error {
+			if u > math.MaxInt {
+				return xerrors.Errorf("too many events")
+			}
+			if err := evt.UnmarshalCBOR(bytes.NewReader(deferred.Raw)); err != nil {
+				return err
+			}
+
+			cpy := evt
+			ems[i].evs[int(u)] = cpy
+			return nil
+		})
+
+		if err != nil {
+			return nil, xerrors.Errorf("error iterating over events for message %d: %w", i, err)
+		}
+
+	}
+
+	return ems, nil
+}
+
+// checkTipsetIndexedStatus verifies if a specific tipset is indexed based on the EventFilter.
+// It returns nil if the tipset is indexed, ErrNotFound if it's not indexed or not specified,
+func (si *SqliteIndexer) checkTipsetIndexedStatus(ctx context.Context, f *EventFilter) error {
+	var tipsetKeyCid []byte
+	var err error
+
+	// Determine the tipset to check based on the filter
+	switch {
+	case f.TipsetCid != cid.Undef:
+		tipsetKeyCid = f.TipsetCid.Bytes()
+	case f.MinHeight >= 0 && f.MinHeight == f.MaxHeight:
+		tipsetKeyCid, err = si.getTipsetKeyCidByHeight(ctx, f.MinHeight)
+		if err != nil {
+			return xerrors.Errorf("failed to get tipset key cid by height: %w", err)
+		}
+	default:
```

Comment:
could we check the min and max tipsets here? or could we do a new query that does a count(*) on the epochs between min and max and confirms that it's equal to the delta?

Reply:
@rvagg It's really tricky to get this right. Even if you verify for the min and max, there might be problems with some tipsets in the range other than min and max which might lead you to erroneously report `ErrNotFound` or erroneously report "no error".

The most common use case of the Events Index is to get events at a specific tipset/epoch and we already handle that well here.

Reply:
@rvagg Also, while I was at this, I also added handling for the "null round at epoch case" i.e. if a user requested epoch is a "null round" on the canonical chain -> simply return empty events and no error.